modules/http4s/src/smithy4s/http4s/internals/SmithyHttp4sServerEndpoint.scala

```diff
@@ -196,18 +198,23 @@ private[http4s] class SmithyHttp4sServerEndpointImpl[F[_], Op[_, _, _, _, _], I,
         .map { case (k, v) => k -> v.map(_._2).toList }
     ).pure[F]
 
-  private def successResponse(output: O): F[Response[F]] = {
+  private def successResponse(output: O): Response[F] = {
     val outputMetadata = outputMetadataEncoder.encode(output)
     val outputHeaders = toHeaders(outputMetadata.headers)
     val statusCode = outputMetadata.statusCode.getOrElse(httpEndpoint.code)
     val httpStatus = status(statusCode)
 
-    putHeaders(Response[F](httpStatus), outputHeaders)
-      .withEntity(output)
-      .pure[F]
+    val baseResponse = putHeaders(Response[F](httpStatus), outputHeaders)
+    if (isEmpty(outputSchema)) baseResponse.withEmptyBody
+    else baseResponse.withEntity(output)
   }
 
-  def compileErrorable(errorable: Errorable[E]): E => Response[F] = {
+  private def isEmpty[A](schema: Schema[A]): Boolean = schema match {
+    case Schema.PrimitiveSchema(_, _, Primitive.PUnit) => true
+    case _                                             => false
+  }
```

Comment:
these changes should move to the `EntityCompiler`

Reply:
did just that